File: lee/src/main/java/org/example/lee/题目/其它/a_多数元素.go
package 其它

// 思路 每个不同元素互相抵消 一定是超过一半的元素留下
func majorityElement(nums []int) int {
	var awe, help int
	for _, num := range nums {
		if help == 0 {
			awe = num
		}
		if awe == num {
			help++
		} else {
			help--
		}
	}
	return awe
}

File: lee/src/main/java/org/example/lee/题目/字符串/a_多数元素.java
package org.example.lee.题目.字符串;

public class a_多数元素 {

	/**
	 * majority elemen 摩尔投票法
	 *
	 * @param nums nums
	 * @return int
	 */
	public int majorityElement(int[] nums) {
		int answer = 0, key = 0;
		for (int num : nums) {
			if (key == 0) {
				answer = num;
			}
			key += answer == num ? 1 : -1;
		}
		return answer;
	}
}
